### PR TITLE
Re-work the DNS part of the test for #26762

### PR DESCRIPTION
**Pull request by mml**
- Dropped the regex test and just test for nslookup exiting 0.
- Moved more setup into BeforeEach and used nested Context for non-local
  case.
- Poll inside the container using a bash loop.
- Aim for less console noise unless something goes wrong.
- Commented out the tests trying to verify that a DNS name is absent.

### CODE REVIEW

File: test/e2e/federated-service.go [hunk 1]
```diff
@@ -148,49 +144,67 @@ var _ = framework.KubeDescribe("Service [Feature:Federation]", func() {
...
+			BeforeEach(func() {
+				framework.SkipUnlessFederated(f.Client)
+
+				// Delete a federated service shard in the default e2e Kubernetes cluster.
+				err := f.Clientset_1_3.Core().Services(f.Namespace.Name).Delete(FederatedServiceName, &api.DeleteOptions{})
+				Expect(err).NotTo(HaveOccurred())
+				waitForFederatedServiceShard(f.Clientset_1_3, f.Namespace.Name, nil, 0)
+			})
+
+			It("should be able to discover a non-local federated service", func() {
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: I think you accidentally left in the 'F'?

Comment (quinton-hoole) [inline test/e2e/federated-service.go]: Heh!  You beat me to it :-)

File: test/e2e/federated-service.go [hunk 2]
```diff
@@ -148,49 +144,67 @@ var _ = framework.KubeDescribe("Service [Feature:Federation]", func() {
...
+				framework.SkipUnlessFederated(f.Client)
+
+				svcDNSNames := []string{
+					fmt.Sprintf("%s.%s.%s", FederatedServiceName, f.Namespace.Name, federationName),
+					fmt.Sprintf("%s.%s.%s.svc.cluster.local.", FederatedServiceName, f.Namespace.Name, federationName),
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: There are a few problems here, as discussed in person:

1. Deleting the service from the local cluster won't actually work, as the Federation Service Controller will re-create it (or at least it should, and if it doesn't, that's a bug).  Instead you'll have to remove all of the endpoints from the service (by deleting all pods that match the service's selector).  That should remove the DNS entry for the local shard of the federated service.  Again, if it doesn't, then that's a bug.
2. Even once the DNS records have been deleted,  DNS propagation will take at least TTL seconds (currently set to 180 in the Federated Service Controller) to propagate.  To get around that you'd probably need to resolve before deleting the service's backends, and then after deleting, until the IP address changes (which might take up to 180 seconds, due to DNS caching).  We might consider reducing that TTL in e2e test clusters to speed up the tests.

File: test/e2e/federated-service.go [hunk 3]
```diff
@@ -148,49 +144,67 @@ var _ = framework.KubeDescribe("Service [Feature:Federation]", func() {
...
+				}
+				for _, name := range svcDNSNames {
+					discoverService(f, name)
+				}
+
+				// TODO(mml): Unclear how to make this meaningful and not terribly
+				// slow.  How long (how many minutes?) do we verify that a given DNS
+				// lookup *doesn't* work before we call it a success?  For now,
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: As above, try for DNS TTL seconds (currently configured to180).

File: test/e2e/federated-service.go [hunk 4]
```diff
@@ -260,17 +275,16 @@ func createService(fcs *federation_release_1_3.Clientset, clusterClientSets []*r
...
 	Expect(err).NotTo(HaveOccurred())
 	for _, cs := range clusterClientSets {
 		waitForFederatedServiceShard(cs, namespace, nservice, 1)
 	}
 }
 
-func discoverService(f *framework.Framework, name string, exists bool) {
-	command := []string{"nslookup", name}
+func discoverService(f *framework.Framework, name string) {
+	command := []string{"sh", "-c", fmt.Sprintf("until nslookup '%s'; do sleep 1; done", name)}
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: I think that you have a potentially endless loop here, right?

Comment (quinton-hoole) [inline test/e2e/federated-service.go]: nit: Didn't you decide that "sh" -"c" was the wrong way to do this?

File: test/e2e/federated-service.go [hunk 5]
```diff
@@ -260,17 +275,16 @@ func createService(fcs *federation_release_1_3.Clientset, clusterClientSets []*r
...
 
-	framework.Logf("Looking for the service with pod command %q", command)
+	defer f.Client.Pods(f.Namespace.Name).Delete(FederatedServicePod, api.NewDeleteOptions(0))
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: Probably better to do this only if the pod creation succeeded below.

File: test/e2e/federated-service.go [hunk 6]
```diff
@@ -285,16 +299,52 @@ func discoverService(f *framework.Framework, name string, exists bool) {
...
-		f.TestContainerOutputRegexp("federated service discovery", pod, 0, []string{
-			`Name:\s+` + FederatedDNS1123Regexp + `\nAddress \d+:\s+` + FederatedIPAddrRegexp + `\s+` + FederatedDNS1123Regexp,
-		})
-	} else {
-		f.TestContainerOutputRegexp("federated service discovery", pod, 0, []string{
-			`nslookup: can't resolve '` + name + `'`,
-		})
+
+	_, err := f.Client.Pods(f.Namespace.Name).Create(pod)
+	Expect(err).
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: Prefer this on one line?

File: test/e2e/federated-service.go [hunk 7]
```diff
@@ -285,16 +299,52 @@ func discoverService(f *framework.Framework, name string, exists bool) {
...
+		NotTo(HaveOccurred(), "Trying to create pod to run %q", command)
+
+	// If we ever get any container logs, stash them here.
+	logs := ""
+
+	logerr := func(err error) error {
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: Perhaps neater and more re-usable to factor this out into a separate function that others can use?

File: test/e2e/federated-service.go [hunk 8]
```diff
@@ -285,16 +299,52 @@ func discoverService(f *framework.Framework, name string, exists bool) {
...
+		}
+		if logs == "" {
+			return err
+		}
+		return fmt.Errorf("%s (%v)", logs, err)
 	}
+
+	// TODO(mml): Eventually check the IP address is correct, too.
+	Eventually(func() error {
+		pod, err := f.Client.Pods(f.Namespace.Name).Get(FederatedServicePod)
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: nit: Similar comment to above.  Perhaps better as a standalone function?

**Comment by quinton-hoole** [comment]
As discussed, this is strictly better than what it replaces, so I'm happy to merge as is to unblock further testing.  Addressing comments can come in a followup PR.

File: test/e2e/federated-service.go (hunk 2 above)
Comment (mml) [inline test/e2e/federated-service.go]: Bug filed: #27623

File: test/e2e/federated-service.go (hunk 3 above)
Comment (mml) [inline test/e2e/federated-service.go]: I can do that, but it adds 9 minutes to every run when things are passing.  I guess I can leave a TODO saying consider dialing down the TTL and consider parallelizing the pod execution.  I already mentioned the latter elsewhere.

File: test/e2e/federated-service.go (hunk 4 above)
Comment (mml) [inline test/e2e/federated-service.go]: It's the write way as long as the next argument is a single string representing a bash command.  Before we had "sh" "-c" "nslookup" "foo", and when you do that, the "foo" goes nowhere.

There is an endless loop here if nslookup never succeeds.  But the Eventually has a timeout, and when we reach that timeout, the test fails and we'll delete the pod on the way out of the test.

File: test/e2e/federated-service.go (hunk 7 above)
Comment (mml) [inline test/e2e/federated-service.go]: I agree, but I'm not sure the final form the test is going to take.  E.g., adding back in the "DNS record missing" assertions, possibly checking the output for the correct IP address.  When it feels solidified, I will go back and extract stuff.

File: test/e2e/federated-service.go (hunk 4 above)
Comment (mml) [inline test/e2e/federated-service.go]: > write way

:laughing:

File: test/e2e/federated-service.go (hunk 3 above)
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: Yeah, I'm happy to leave this test disabled until we've dialed down the TTL. Just wanted to explain what the solution would be :-)

File: test/e2e/federated-service.go (hunk 4 above)
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: Agreed.  Just wanted to check.  Best check what the timeout on Eventually is.  I'm not sure offhand.

File: test/e2e/federated-service.go (hunk 7 above)
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: Fair enough.

**Pull request merged**